util/bloom_test.cc

```diff
@@ -601,6 +603,255 @@ TEST_P(FullBloomTest, OptimizeForMemory) {
   }
 }
 
+TEST_P(FullBloomTest, ReserveBloomAndRibbonFilterConstructionMemory) {
+  const BloomFilterPolicy::Mode mode = GetParam();
+  const bool can_reserve_memory =
+      (mode == BloomFilterPolicy::Mode::kFastLocalBloom) ||
+      (mode == BloomFilterPolicy::Mode::kStandard128Ribbon);
+  constexpr std::size_t kSizeDummyEntry = 256 * 1024;
+  constexpr std::size_t kCacheCapacity = 8 * 1024 * 1024;
+  std::size_t memory_reserved = 0;
+  std::size_t num_dummy_entries = 0;
+
+  BlockBasedTableOptions table_options;
+  LRUCacheOptions lo;
+  lo.capacity = kCacheCapacity;
+  lo.num_shard_bits = 0;  // 2^0 shard
+  lo.strict_capacity_limit = true;
+  std::shared_ptr<Cache> cache(NewLRUCache(lo));
+  table_options.block_cache = cache;
+  table_options.filter_policy.reset(
+      new BloomFilterPolicy(FLAGS_bits_per_key, mode));
+  table_options.reserve_table_builder_memory = true;
+
+  FilterBuildingContext ctx(table_options);
+
+  ASSERT_EQ(cache->GetPinnedUsage(), 0 * kSizeDummyEntry);
+  std::unique_ptr<FilterBitsBuilder> filter_bits_builder(
+      table_options.filter_policy->GetBuilderWithContext(ctx));
+
+  const std::size_t num_entries = 10000;
+  char key_buffer[sizeof(int)];
+  for (std::size_t i = 0; i < num_entries; ++i) {
+    filter_bits_builder->AddKey(Key(static_cast<int>(i), key_buffer));
+  }
+  // 0xffffffffU is max size supported by implementation for buf length
+  std::unique_ptr<const char[]> buf(new char[0xffffffffU]);
+  Slice filter = filter_bits_builder->Finish(&buf);
+
+  if (can_reserve_memory) {
+    // Memory reserved now: mutable_buffer_'s size (approximated by
+    // filter.size())
+    memory_reserved = filter.size();
+  } else {
+    memory_reserved = 0;
+  }
+  num_dummy_entries = static_cast<std::size_t>(
+      std::ceil(memory_reserved / static_cast<double>(kSizeDummyEntry)));
+  EXPECT_GE(cache->GetPinnedUsage(), num_dummy_entries * kSizeDummyEntry);
+  EXPECT_LT(cache->GetPinnedUsage(), (num_dummy_entries + 1) * kSizeDummyEntry);
+
+  filter_bits_builder.reset();
+  EXPECT_EQ(cache->GetPinnedUsage(), 0 * kSizeDummyEntry);
+}
+
+TEST_P(FullBloomTest, RibbonFilterFallBackOnLargeBanding) {
```

Comment:
I agree this fallback case is worth testing. I'm ok with this one being tested at the unit (FilterBitsBuilder) level because it is a little trickier (but doable) at the integration level, and other tests should give us confidence in the integration logic (which this fallback case does not complicate).